src/ucc/api/ucc.h

```diff
@@ -635,15 +635,18 @@ enum ucc_context_attr_field {
  *
  * @brief OOB collective operation for creating the context
  */
-typedef struct ucc_context_oob_coll {
+typedef struct ucc_oob_coll {
     ucc_status_t    (*allgather)(void *src_buf, void *recv_buf, size_t size,
                                  void *allgather_info,  void **request);
     ucc_status_t    (*req_test)(void *request);
     ucc_status_t    (*req_free)(void *request);
     uint32_t        participants;
     void            *coll_info;
-}  ucc_context_oob_coll_t;
+    uint32_t        rank;
+}  ucc_oob_coll_t;
```

Comment:
- Naming - How about rank -> oob_ep ? Since there is a relation between ep (of team) and the “rank” here. 
- Also, participants -> num_oob_eps ?
- We need to add a constraint saying that “rank” is unique among the participants.
- We need to add a constraint saying that if a certain oob is used for context create, the same oob should be used for team create.
- We need to define the relation between “rank” and team’s “ep”, when we create the team.
- What would be a relation with “rank” and context’s ctx_id ? No relation?

Reply:
> * Naming - How about rank -> oob_ep ? Since there is a relation between ep (of team) and the “rank” here.
Done
> * Also, participants -> num_oob_eps ?
Done
> * We need to add a constraint saying that “rank” is unique among the participants.
Added desc
> * We need to add a constraint saying that if a certain oob is used for context create, the same oob should be used for team create.
Why? team oob is general is different from ctx oob (e.g., team oob can be a subset of ctx oob). We wanted to add another constraint: if team is created from multiple "GLOBAL" contexts then ALL those ctxs must have same oob. Is that what you mean here?
> * We need to define the relation between “rank” and team’s “ep”, when we create the team.
This is most interesting one. Basically, these are now 2 ways how user/runtime can pass "the rank of a process in a team" down to ucc_team_create_post: 1st way - EP_CONTIG + ep as input int parameter; 2nd way as team_params.oob.oob_ep. 
I think we should allow all combinations:
1. None is provided. No EP/EP_RANGE_CONTIG, no OOB. Then we compute EP internally and return as output.
2. Either one is provided: either EP w/o OOB (team oob is not necessary if ctx has oob and ep_map is provided for team_create) ; OR OOB is provided w/o EP - then EP is taken from oob.oob_ep
3. Both are provided. Then we demand that the input is consistent, ie oob.oob_ep == EP contig value.
4. If value of EP is provided and it is NOT contig, then it will be used as EP for the API (but it is up to internal implementation to utilize oob_ep if OOB is provided).
@manjugv does it make sense?
> * What would be a relation with “rank” and context’s ctx_id ? No relation?
Yes, no relation.